src/protocol/loli_nrf24l01.c

```diff
@@ -0,0 +1,420 @@
+                NRF24L01_ReadPayload(packet, LOLI_PACKET_SIZE);
+                if (packet[0] == 'O' && packet[1] == 'K') {
+                    NRF24L01_WriteRegisterMulti(NRF24L01_0A_RX_ADDR_P0, rx_tx_addr, 5);
+                    NRF24L01_WriteRegisterMulti(NRF24L01_10_TX_ADDR, rx_tx_addr, 5);
+                    PROTOCOL_SetBindState(0);
+                    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);
+                    NRF24L01_FlushRx();
+                    phase = DATA1;
+                    break;
+                }
+            }
+            count++;
+            if (count > 50) {
+                phase = BIND1;
+            }
+            delay = 1000;
+            break;
+        case DATA1:
+            // got a telemetry packet ?
+            if (NRF24L01_ReadReg(NRF24L01_07_STATUS) & BV(NRF24L01_07_RX_DR)) {  // RX fifo data ready
+                NRF24L01_ReadPayload(packet, LOLI_PACKET_SIZE);
+                update_telemetry();
+            }
+            if (rxConfigChanged()) {
+                count = 0;
+                phase = SET_RX_CONFIG;
+                break;
+            }
+            NRF24L01_SetTxRxMode(TXRX_OFF);
+            NRF24L01_SetTxRxMode(TX_EN);
+            NRF24L01_WriteReg(NRF24L01_00_CONFIG, 0x0a);  // 8bit CRC, TX
+            // send data packet
+            send_packet(0);
+            phase = DATA2;
+            delay = 2000;
+            break;
+        case DATA2:
+            // switch to RX mode
+            NRF24L01_SetTxRxMode(TXRX_OFF);
+            NRF24L01_FlushRx();
+            NRF24L01_SetTxRxMode(RX_EN);
+            NRF24L01_WriteReg(NRF24L01_00_CONFIG, 0x3b);  // 8bit CRC, RX
+            phase = DATA1;
+            delay = 18000;
+            break;
+        case SET_RX_CONFIG:
+            NRF24L01_SetTxRxMode(TXRX_OFF);
+            NRF24L01_SetTxRxMode(TX_EN);
+            NRF24L01_WriteReg(NRF24L01_00_CONFIG, 0x0a);  // 8bit CRC, TX
+            send_rx_config();
+            if (++count >= 15) {
+                phase = DATA1;
+            }
+            delay = 20000;
+            break;
+    }
+    return delay;
+}
+
+static void init_txid()
+{
+    u32 lfsr = 0xb2c54a2ful;
+#ifndef USE_FIXED_MFGID
+    u8 var[12];
+    MCU_SerialNumber(var, 12);
+    for (int i = 0; i < 12; ++i) {
+        rand32_r(&lfsr, var[i]);
+    }
+#endif
+    if (Model.fixed_id) {
+       for (u8 i = 0, j = 0; i < sizeof(Model.fixed_id); ++i, j += 8)
+           rand32_r(&lfsr, (Model.fixed_id >> j) & 0xff);
+    }
+    // Pump zero bytes for LFSR to diverge more
+    for (u8 i = 0; i < sizeof(lfsr); ++i) rand32_r(&lfsr, 0);
+
+    // tx id
+    rx_tx_addr[0] = (lfsr >> 24) & 0xFF;
+    rx_tx_addr[1] = ((lfsr >> 16) & 0xFF) % 0x30;
+    rx_tx_addr[2] = (lfsr >> 8) & 0xFF;
+    rx_tx_addr[3] = lfsr & 0xFF;
+    rand32_r(&lfsr, 0);
+    rx_tx_addr[4] = (lfsr >> 24) & 0xFF;
+
+    // rf channels for data phase (manually set on "original" TX ?)
+    for (u8 i=0; i < LOLI_NUM_CHANNELS; i++) {
```

Comment:
original Tx use channel RSSI scan to find the best set of channels in the begining. can we do the same?

Reply:
NRF24L01 doesn't have actual RSSI (only a Carrier Detect flag) and I don't think that's useful, eg we may find a frequency is "free" during bind, but nothing guaranty it will still be free tomorrow. But I'll modify that so they're better spread across the spectrum.

Reply:
the protocol supports dynamic switch the hop freq table. Tx can proactive to ask Rx to use another hop table if it want to.

I am not sure what's common practice in other protocols. your call for sure.

Reply:
except for a few protocols that use random channels at each startup, most protocols always use the same channels, derived from a semi-arbitrary transmitter ID.

Reply:
To my understanding the "original" transmitter doesn't check RSSI but let the user manually selects 5 channels (same for nrf24 address). Perhaps I'm wrong, code of the original transmitter isn't easy to read ;)